File: backend-ts/src/connections/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto';

/**
 * AES-256-GCM token-at-rest encryption (DA-12) — CROSS-BACKEND INTERCHANGEABLE.
 *
 * Byte-compatible with `backend-python/app/connections/crypto.py`: both backends
 * read/write the SAME `plaid_items.access_token` BYTEA column, so the on-disk
 * layout MUST be identical:
 *
 *   nonce(12 bytes) || ciphertext(N) || tag(16 bytes)
 *
 * A token written by FastAPI decrypts here and vice-versa (proven by a parity
 * test). Node's `aes-256-gcm` exposes the tag via `getAuthTag()`; we append it
 * last so the layout matches Python's `AESGCM.encrypt` (which returns
 * `ciphertext || tag`). The key is the base64-decoded `APP_ENCRYPTION_KEY`
 * (exactly 32 bytes). No plaintext token is ever logged (DA-14).
 */

export const NONCE_BYTES = 12;
export const TAG_BYTES = 16;

export class TokenCipherError extends Error {}

/** Decode + validate the base64 AES-256 key (must be exactly 32 bytes). */
function loadKey(appEncryptionKey: string): Buffer {
  if (!appEncryptionKey) {
    throw new TokenCipherError('APP_ENCRYPTION_KEY is not configured.');
  }
  let key: Buffer;
  try {
    key = Buffer.from(appEncryptionKey, 'base64');
  } catch {
    throw new TokenCipherError('APP_ENCRYPTION_KEY is not valid base64.');
  }
  if (key.length !== 32) {
    throw new TokenCipherError(
      'APP_ENCRYPTION_KEY must decode to 32 bytes (AES-256).',
    );
  }
  return key;
}

/** Encrypt `plaintext` -> `nonce(12) || ciphertext || tag(16)` bytes. */
export function encryptToken(
  plaintext: string,
  appEncryptionKey: string,
): Buffer {
  const key = loadKey(appEncryptionKey);
  const nonce = randomBytes(NONCE_BYTES);
  const cipher = createCipheriv('aes-256-gcm', key, nonce);
  const ciphertext = Buffer.concat([
    cipher.update(Buffer.from(plaintext, 'utf8')),
    cipher.final(),
  ]);
  const tag = cipher.getAuthTag(); // 16 bytes
  return Buffer.concat([nonce, ciphertext, tag]);
}

/** Decrypt a `nonce(12) || ciphertext || tag(16)` blob back to the token. */
export function decryptToken(blob: Buffer, appEncryptionKey: string): string {
  const key = loadKey(appEncryptionKey);
  if (blob.length < NONCE_BYTES + TAG_BYTES) {
    throw new TokenCipherError('Ciphertext blob is too short to be valid.');
  }
  const nonce = blob.subarray(0, NONCE_BYTES);
  const tag = blob.subarray(blob.length - TAG_BYTES);
  const ciphertext = blob.subarray(NONCE_BYTES, blob.length - TAG_BYTES);
  const decipher = createDecipheriv('aes-256-gcm', key, nonce);
  decipher.setAuthTag(tag);
  try {
    return Buffer.concat([
      decipher.update(ciphertext),
      decipher.final(),
    ]).toString('utf8');
  } catch {
    // Never leak why (bad tag / wrong key / corrupt blob).
    throw new TokenCipherError('Token decryption failed.');
  }
}

File: backend-python/app/connections/crypto.py
"""AES-256-GCM token-at-rest encryption (DA-12) — CROSS-BACKEND INTERCHANGEABLE.

A Plaid ``access_token`` is secret. It is stored only as ciphertext in
``plaid_items.access_token`` (``BYTEA``), never as plaintext.

The on-disk format is a fixed, backend-neutral byte layout so a token written by
FastAPI decrypts in NestJS and vice-versa (both read/write the SAME row)::

    nonce(12 bytes) || ciphertext(N) || tag(16 bytes)

* Key: the base64-decoded ``APP_ENCRYPTION_KEY`` (exactly 32 bytes = AES-256).
* Nonce: 12 random bytes per write (GCM standard), prepended.
* Tag: GCM's 16-byte authentication tag, appended (this is what Python's
  ``AESGCM`` returns concatenated to the ciphertext, and what Node's
  ``aes-256-gcm`` exposes via ``getAuthTag()`` — packing it last makes both
  byte-identical).

No plaintext token is ever logged or returned (DA-14).
"""

from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

NONCE_BYTES = 12
TAG_BYTES = 16


class TokenCipherError(Exception):
    """Raised when the encryption key is missing/invalid or a blob won't decrypt."""


def _load_key(app_encryption_key: str) -> bytes:
    """Decode + validate the base64 AES-256 key (must be exactly 32 bytes)."""
    if not app_encryption_key:
        raise TokenCipherError("APP_ENCRYPTION_KEY is not configured.")
    try:
        key = base64.b64decode(app_encryption_key, validate=True)
    except (ValueError, base64.binascii.Error) as exc:  # type: ignore[attr-defined]
        raise TokenCipherError("APP_ENCRYPTION_KEY is not valid base64.") from exc
    if len(key) != 32:
        raise TokenCipherError("APP_ENCRYPTION_KEY must decode to 32 bytes (AES-256).")
    return key


def encrypt_token(plaintext: str, app_encryption_key: str) -> bytes:
    """Encrypt ``plaintext`` -> ``nonce(12) || ciphertext || tag(16)`` bytes.

    A fresh random nonce is generated per call. ``AESGCM.encrypt`` returns
    ``ciphertext || tag`` already, so we simply prepend the nonce.
    """
    key = _load_key(app_encryption_key)
    nonce = os.urandom(NONCE_BYTES)
    aesgcm = AESGCM(key)
    ciphertext_and_tag = aesgcm.encrypt(nonce, plaintext.encode("utf-8"), None)
    return nonce + ciphertext_and_tag


def decrypt_token(blob: bytes, app_encryption_key: str) -> str:
    """Decrypt a ``nonce(12) || ciphertext || tag(16)`` blob back to the token."""
    key = _load_key(app_encryption_key)
    if len(blob) < NONCE_BYTES + TAG_BYTES:
        raise TokenCipherError("Ciphertext blob is too short to be valid.")
    nonce = blob[:NONCE_BYTES]
    ciphertext_and_tag = blob[NONCE_BYTES:]
    aesgcm = AESGCM(key)
    try:
        plaintext = aesgcm.decrypt(nonce, ciphertext_and_tag, None)
    except Exception as exc:  # InvalidTag etc. — never leak detail.
        raise TokenCipherError("Token decryption failed.") from exc
    return plaintext.decode("utf-8")
